### PR TITLE
Add compatibility hacks for Adreno 3xx

### PR DESCRIPTION
**Pull request by MednauN**
As mentioned in many issues (https://github.com/godotengine/godot/issues/12816 https://github.com/godotengine/godot/issues/14771 https://github.com/godotengine/godot/issues/11693, maybe some more), Godot doesn't run on devices with Adreno 3xx GPU.
I tried to fix this... and found many strange bugs. For some of these bugs I wrote hacks and finally I managed to launch my game on ASUS Z010D (with Adreno 306). Also tested on Sony G8441 with Adreno 540, but the later one doesn't have any issues to begin with.
So, here is what I found out:
**Issue 1**
Shader compiler hangs on compilation of `pow` function (yes, I'm 100% sure, I checked, double-checked and triple-checked this). That's mostly why 3D scenes hang. My guess, the compiler tries to inline `pow`, but something goes wrong and it fails. Surprising fix for this bug is to make a custom function which calls `pow` and call that function instead. Pretty much, like this:
```
float pow_no_inline(float x, float y) {
	return pow(x, y);
}
#define pow(x, y) pow_no_inline(x, y)
```

**Issue 2**
Now shaders compile, but it takes ~3.5 sec to compile one shader. Woops.
And that's because of `OmniLightData`, `SpotLightData` amd `ReflectionProbeData` structures. They contain arrays with length of max possible count of lights/reflections and that's about 400 for Adreno 306. Probably, compiler tries to unfold `for` cycles with hypothetically 400+ iterations and it takes quite a lot of time. After setting max lights and max reflections to smaller value, shader compiler starts working really fast.
In the PR I just moved `max_lights` and `max_reflections` to the project settings.

**Issue 3**
Next bug I found was crashing on shader compilation. For example, this shader produces 100% crash:
```
shader_type canvas_item;

uniform sampler2D mask_texture: hint_white;
uniform bool grayscale;

void fragment() {
	vec4 tex_color = texture(TEXTURE, UV);
	COLOR = vec4(tex_color.rgb, texture(mask_texture, UV).r * tex_color.a);
	
	if (grayscale)
	{
		COLOR.rgb = vec3(dot(COLOR.rgb, vec3(0.3, 0.59, 0.11)));
	}
	
}
```
And the line that actually causes crash is `if (grayscale)`.
Adreno 306 doesn't like uniform bool variables. Strangely enough it only happens to variables from the user code. And simply replacing condition with `if (int(grayscale) != 0)` eliminates the crash.
My final solution is to wrap all user-defined boolean variables into a macro defined like this:
```
#define UNWRAP_BOOL(b) (int(b) != 0)
```
And no more crash with this.

**Now the issues I found, but didn't solve:**
**Issue 4**
Vertex lighting doesn't work. Meshes are rendered distorted or not rendered at all. No errors in logcat. I don't use vertex lighting in my project, so I didn't research, why this happens and how to fix it.
I found out that after commenting out this line:
https://github.com/godotengine/godot/blob/1fa9aac3e415f53a095e955c8a37000629d56dde/drivers/gles3/shaders/scene.glsl#L171
meshes are rendered normally (but unlit) and that's where I stopped.

**Issue 5**
Using `SCREEN_TEXTURE` in the shader leads to a 100% GPU deadlock. The logcat message is like this:
`<gsl_ldd_control:427>: ioctl fd XX code 0x400c0907 (IOCTL_KGSL_DEVICE_WAITTIMESTAMP_CTXTID) failed: errno 35 Resource deadlock would occur`
I don't have a fix for this one too.
Looks like deadlock occur on calls of `scene_render->state.effect_blur_shader.bind();` in `RasterizerCanvasGLES3::_copy_texscreen`. At least, no deadlock after commenting out these calls.

**Issue 6**
Showing over 3D scene some UI with `ViewportContainer` (which contains another 3D scene) screws up the render of the main 3D scene. Even after deleting `ViewportContainer`, the main scene doesn't recover. Scene inside container is rendered just fine.
No fix for this, ~but I'm still working on it, maybe I can find one.~ UPD: won't fix this bug, screw it.

That's all I found so far.
TL;DR:
1. You can launch on Adreno 3xx now
2. Do not use vertex lighting
3. Do not use SCREEN_TEXTURE
4. Do not use ViewportContainer
5. Or better wait for GLES 2.0 renderer
5. F.. you, Qualcomm

### CODE REVIEW

**Comment by reduz** [comment]
Adreno is the worst piece of crap GPU ever invented. I even get random crashes and deadlocks on more modern versions for really strange reasons. I'm not sure if fixing this is worth it, as karroffel will soon get the GLES2 backend to work..

**Comment by mhilbrunner** [comment]
Couldn't `in the PR I just moved max_lights and max_reflections to the project settings` be a useful thing to have in general?

**Comment by akien-mga** [comment]
Thanks a lot for your contribution!
We are now entering a strict release freeze for Godot 3.0 and will only consider major bug fixes. We won't merge new features and enhancements anymore until 3.0 is released.

Moving this PR to the 3.1 milestone, to be reviewed once the release freeze is lifted. It could eventually be cherry-picked for a future 3.0.1 maintenance release if it doesn't change the user-facing APIs and doesn't compromise the engine's stability.

**Comment by reduz** [comment]
Lets wait for 3.1 and GLES2 backend to be properly implemented

**Comment by ret80** [comment]
@reduz
> Lets wait for 3.1 and GLES2 backend to be properly implemented

At the moment we have a stable version of the engine that does not work on half of the devices with the GPU adreno and the best recommendation is that it will wait for version 3.1 which will appear unknown when!
I propose to fix this as soon as possible with the help of hacks (let it not be the best solution but at least it will work).

**Comment by Zireael07** [comment]
@endragor says this PR reduces the times spent on compiling shaders. Can the part that reduces the times spent on shader compiling be isolated and PR'ed anew (separately from the Adreno stuff)?

**Comment by Megalomaniak** [comment]
@reduz ah but this doesn't only benefit the adreno. Some of the things here are overall improvements that probably should be separated into their own PR's and then merged perhaps. At least the solution to Issue 2.

**Comment by jknightdoeswork** [comment]
>Lets wait for 3.1 and GLES2 backend to be properly implemented

Ok. I think it's time to merge this.

**Comment by Calinou** [comment]
@jknightdoeswork What @reduz meant here is that the now-implemented GLES2 renderer is considered to replace the GLES3 renderer on low-end platforms like Android. This is because GLES3 support is often too buggy to be usable on mobile platforms.

**Comment by jknightdoeswork** [comment]
@Calinou I'm specifically concerned with this issue in the html5 export: https://github.com/godotengine/godot/issues/21957

This aspect of this pull request makes a lot of sense to me:
>In the PR I just moved max_lights and max_reflections to the project settings.

I don't imagine that the 4.0 vulkan release will have any positive effects on the 3d performance of the html export.

**Comment by clayjohn** [comment]
@jknightdoeswork Towards the end of the issue you linked you can see that it is indeed resolved by using GLES2 instead of GLES3. The user reports load times of 3-4 seconds after switching to GLES2. 

> I don't imagine that the 4.0 vulkan release will have any positive effects on the 3d performance of the html export.

You are probably right. Which will make using GLES2 for web even more important.

**Comment by jknightdoeswork** [comment]
My scenes have many lighting bugs on GLES2. Also, arent we actually talking
about WebGL1 vs WebGL2? Not actually GLES2 and GLES3?

**Comment by clayjohn** [comment]
@jknightdoeswork The GLES2 backend exports to WebGL1 and the GLES3 backend exports to WebGL2. 

The preferred approach is still to use the GLES2 backend and export to WebGL1. If you have bugs in GLES2 you should report them so that someone can fix them. That being said, I made a PR that allows you to change the max lights and max reflections in the project settings :)

**Comment by jknightdoeswork** [comment]
@clayjohn Can you link it?

**Comment by akien-mga** [comment]
It's right above:
Screenshot_20190730_183900
